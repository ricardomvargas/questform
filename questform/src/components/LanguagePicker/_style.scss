%menu-item-options-default {
  background-color: $blue-white;
  border: solid 2px $blue-gray;
  border-radius: 20px;
  position: absolute;
  width: 45px;
  padding: 2px;
  top: 32px;
}
.lang-menu {
  .img-flag {
    width: 20px;
    height: 20px;
  }
  .img-options {
    width: 12px;
    height: 12px;
    margin-left: 3px;
  }
  .display-langs {
    border: none;
    background: none;
    padding: 0;
    margin: 0;
  }
  &-options {
    @extend %menu-item-options-default;
    display: flex;
    justify-content: space-between;
  }
  &-options-hide {
    @extend %menu-item-options-default;
    display: none;
  }
}
