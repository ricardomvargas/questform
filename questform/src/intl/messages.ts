import appMessagesEN from './app/en';
import appMessagesPT from './app/pt';
import dashboardEN from './pages/dashboard/en';
import dashboardPT from './pages/dashboard/pt';
import questionEN from './pages/question/en';
import questionPT from './pages/question/pt';
import surveyEN from './pages/survey/en';
import surveyPT from './pages/survey/pt';

export default {
  en: {
    ...appMessagesEN,
    ...dashboardEN,
    ...questionEN,
    ...surveyEN,
  },
  pt: {
    ...appMessagesPT,
    ...dashboardPT,
    ...questionPT,
    ...surveyPT,
  },
};
