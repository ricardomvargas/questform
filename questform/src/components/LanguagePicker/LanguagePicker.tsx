import React, { useState, useContext } from 'react';
import { useIntl } from 'react-intl';

import LangContext from '../../context';

import brFlagRounded from '../../assets/images/br-flag-rounded.png';
import ukFlagRounded from '../../assets/images/uk-flag-rounded.png';
import arrowDownImg from '../../assets/images/arrow-down.png';
import closeImg from '../../assets/images/close.png';

import { storeLanguage, appLangs } from '../../intl/languageStorage';

const LanguagePicker = () => {
  const intl = useIntl();
  const { updateAppLang, appLang } = useContext(LangContext);
  const [langOptionsVisible, setLangOptionsVisible] = useState(false);
  const [langOptionsLeftPosition, setlangOptionsLeftPosition] = useState<number>();

  const getButtonFlagPosition = (el: HTMLImageElement) => {
    let x: number = 0;
    while (el && !isNaN(el.offsetLeft) && !isNaN(el.offsetTop)) {
      x += el.offsetLeft - el.scrollLeft;
      el = el.offsetParent as HTMLImageElement;
    }
    return x - 36;
  };

  const updateMenuLangVisibility = (): void => {
    setlangOptionsLeftPosition(
      getButtonFlagPosition(document.getElementById('btn-show-menu-item') as HTMLImageElement)
    );
    setLangOptionsVisible(!langOptionsVisible);
  };

  const handleLanguageChange = (newLang: string): void => {
    storeLanguage(newLang);
    if (updateAppLang) updateAppLang(newLang);
    updateMenuLangVisibility();
  };

  const closeMenuItemOptions = (): void => setLangOptionsVisible(false);
  // #TODO: Analisar se não é melhor usar um useEffect para registrar esse evento
  window.addEventListener('resize', closeMenuItemOptions);

  return (
    <section className="lang-menu">
      <div className="lang-menu-item">
        <img
          src={appLang === appLangs.english ? ukFlagRounded : brFlagRounded}
          className="img-flag"
          alt={appLang}
          title={appLang}
          onClick={updateMenuLangVisibility}
        />
        <img
          id="btn-show-menu-item"
          className="img-options"
          src={langOptionsVisible ? closeImg : arrowDownImg}
          alt={
            langOptionsVisible
              ? intl.formatMessage({ id: 'buttonClose' })
              : intl.formatMessage({ id: 'buttonOpen' })
          }
          title={
            langOptionsVisible
              ? intl.formatMessage({ id: 'buttonClose' })
              : intl.formatMessage({ id: 'buttonOpen' })
          }
          onClick={updateMenuLangVisibility}
        />
      </div>
      <div
        style={{ left: langOptionsLeftPosition }}
        className={langOptionsVisible ? 'lang-menu-options' : 'lang-menu-options-hide'}
      >
        <img
          src={ukFlagRounded}
          className="img-flag"
          alt={intl.formatMessage({ id: 'changeToEnglish' })}
          title={intl.formatMessage({ id: 'changeToEnglish' })}
          onClick={() => handleLanguageChange(appLangs.english)}
        />
        <img
          src={brFlagRounded}
          className="img-flag"
          alt={intl.formatMessage({ id: 'changeToPortuguese' })}
          title={intl.formatMessage({ id: 'changeToPortuguese' })}
          onClick={() => handleLanguageChange(appLangs.portugueseBR)}
        />
      </div>
    </section>
  );
};

export default LanguagePicker;
