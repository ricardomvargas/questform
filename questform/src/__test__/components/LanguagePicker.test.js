/**
 * @jest-environment jsdom
 */
import React from 'react';
import '@testing-library/jest-dom';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

import LangContext from '../../context';

import LanguagePicker from '../../components/LanguagePicker/LanguagePicker';

import messages from '../../intl/messages';

const updateAppLang = jest.fn();
const appLang = 'en';

const Wraper = ({ children }) => (
  <IntlProvider locale={appLang} messages={messages.en}>
    {children}
  </IntlProvider>
);

test('Check if match with snapshot', () => {
  const header = render(
    <Wraper>
      <LangContext.Provider value={{ updateAppLang, appLang }}>
        <LanguagePicker />
      </LangContext.Provider>
    </Wraper>
  );
  expect(header).toMatchSnapshot();
});
