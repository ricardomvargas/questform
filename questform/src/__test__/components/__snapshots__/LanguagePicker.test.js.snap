// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if match with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="lang-menu"
      >
        <div
          class="lang-menu-item"
        >
          <img
            alt="en"
            class="img-flag"
            src="/assets/images/uk-flag-rounded.png"
            title="en"
          />
          <img
            alt="Open"
            class="img-options"
            id="btn-show-menu-item"
            src="/assets/images/arrow-down.png"
            title="Open"
          />
        </div>
        <div
          class="lang-menu-options-hide"
        >
          <img
            alt="Change to english"
            class="img-flag"
            src="/assets/images/uk-flag-rounded.png"
            title="Change to english"
          />
          <img
            alt="Change to portuguese"
            class="img-flag"
            src="/assets/images/br-flag-rounded.png"
            title="Change to portuguese"
          />
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="lang-menu"
    >
      <div
        class="lang-menu-item"
      >
        <img
          alt="en"
          class="img-flag"
          src="/assets/images/uk-flag-rounded.png"
          title="en"
        />
        <img
          alt="Open"
          class="img-options"
          id="btn-show-menu-item"
          src="/assets/images/arrow-down.png"
          title="Open"
        />
      </div>
      <div
        class="lang-menu-options-hide"
      >
        <img
          alt="Change to english"
          class="img-flag"
          src="/assets/images/uk-flag-rounded.png"
          title="Change to english"
        />
        <img
          alt="Change to portuguese"
          class="img-flag"
          src="/assets/images/br-flag-rounded.png"
          title="Change to portuguese"
        />
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
